File: proyecto/cppcalc/calculator.cpp
#include "calculator.h"
#include "parser.h"
#include "ast.h"
#include <string>
#include <iostream>
#include <sstream>
using namespace std;

Calculator::Calculator():memory(0){
}

int Calculator::eval(string expr){

  Parser* parser = new Parser(new istringstream(expr));

  AST* tree = parser->parse();

  int result = tree->evaluate();

  delete tree;
  delete parser;
  
  return result;
}

void Calculator::store(int val){
  memory =val;
}

int Calculator::recall(){
  return memory;
}

int Calculator::Clean(){
  memory =0;
  return 0;
}

int Calculator::plus(int val){
  memory +=val;
  return memory;
}


int Calculator::minus(int val){
  memory -=val;
  return memory;
}

void Calculator::crearVar(string nombre, int valor){

  string vallor =  static_cast<ostringstream*>(&(ostringstream() << valor))->str();

  salida(nombre, vallor);
  
  if(variables.count(nombre) ==1){
     variables[nombre] =valor;
  }else{
     variables.insert(pair<string,int>(nombre,valor));
  }  
}

int Calculator::getVariables(string nombre){

  if(variables.count(nombre) ==1){
    return variables[nombre];
  }else{
    return 0;
  }  
}

string Calculator::salida(string nombre, string valor){

  if(salidaF == ""){
    salidaF = salidaF + nombre + " <- " + valor;
  }else{
    salidaF += ", " + nombre + " <- " + valor;    
  }
  return salidaF;

}

string Calculator::getSalida(){
  return salidaF;
}

void Calculator::limpiarSalida(){
  salidaF = "";
}

			      

File: proyecto/rubycalc/lib/calculator.rb
require 'parser'
require 'ast'
require 'stringio'

class Calculator  
  attr_accessor :memory
  
  # attr_reader :memory
  # attr_writer :memory
  
  def initialize()
    @memory = 0
    @variables = Array.new
    @nombres = Array.new
    @salidaF = ""
  end
  
  def eval(expr)
    parser = Parser.new(StringIO.new(expr))
    ast = parser.parse()
    return ast.evaluate()
  end

  def crearVar(nombre, valor)    
    entra = true    
    posN = @nombres.length
    
    while posN >= 0 do
      if @nombres[posN] == nombre then
        @variables[posN] = valor
        entra = false
      end
      posN -= 1
    end

    if entra == true then
      @variables.push(valor)
      @nombres.push(nombre)
    end

    salida(nombre,valor)
  end

  def getVariables(nombre)
    posN = @nombres.length
    while posN >= 0 do
      if @nombres[posN] == nombre then
        return @variables[posN]
      end
      posN -= 1
    end
    return 0
  end

  def salida(nombre,valor)
    if @salidaF == "" then
      @salidaF = @salidaF + nombre + " <- " + valor.to_s
    else
      @salidaF = @salidaF + ", " + nombre + " <- " + valor.to_s
    end
  end

  def getSalida()
    return @salidaF
  end

  def limpiarSalida()
    @salidaF = ""    
  end
end
